### PR TITLE
INTYGFV-15083: Implementation of questions related to category Intyget avser.

### PR DESCRIPTION
**Pull request by ChrilleSigerhed**
Added rows & columns to CertificateDataConfigCheckboxMultipleCode.

### CODE REVIEW

File: support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.junit.jupiter.api.Test;
+import se.inera.intyg.common.support.facade.model.CertificateDataElement;
+import se.inera.intyg.common.support.facade.model.value.CertificateDataValueCodeList;
+import se.inera.intyg.common.support.facade.model.value.CertificateDataValueType;
+
+public abstract class ValueCheckboxMultipleCodeTest extends ValueTest {
+
+    protected abstract CertificateDataElement getElementWithValues();
+
+    protected abstract List<Object> getValues();
```
Comment (mww-sopra-steria) [inline support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java:35]: I would suggest that you change to List<CertificateDataValueCode>.

File: support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.List;
+import org.junit.jupiter.api.Test;
+import se.inera.intyg.common.support.facade.model.CertificateDataElement;
+import se.inera.intyg.common.support.facade.model.value.CertificateDataValueCodeList;
+import se.inera.intyg.common.support.facade.model.value.CertificateDataValueType;
+
+public abstract class ValueCheckboxMultipleCodeTest extends ValueTest {
```
Comment (mww-sopra-steria) [inline support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java:31]: Change the name to ValueCodeListTest as it is a testFixture for CODE_LIST and will be reused in multiple places combined with different configs.

File: support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    protected abstract List<Object> getValues();
+
+    @Override
+    protected CertificateDataValueType getType() {
+        return CertificateDataValueType.CODE_LIST;
+    }
+
+    @Test
+    void shouldIncludeCodeValueList() {
```
Comment (mww-sopra-steria) [inline support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java:43]: Maybe could to use the "parameterised"-test pattern? It could contain list of codes, empty list, null list etc. 

(This comment is for both tests)

But I understand the issue with the difference between different type of classes that gets mapped. Maybe can use generics?

File: ts/ts-bas/src/main/java/se/inera/intyg/common/ts_bas/v7/model/converter/InternalToCertificate.java [hunk 4]
```diff
@@ -24,14 +24,25 @@
...
 
     public Certificate convert(TsBasUtlatandeV7 internalCertificate, CertificateTextProvider texts) {
+        int index = 0;
         return CertificateBuilder.create()
             .metadata(MetaDataGrundData.toCertificate(internalCertificate, texts))
+            .addElement(
+                CategoryIntygetAvser.toCertificate(index++, texts)
+            )
+            .addElement(
+                QuestionIntygetAvser.toCertificate(
```
Comment (mww-sopra-steria) [inline ts/ts-bas/src/main/java/se/inera/intyg/common/ts_bas/v7/model/converter/InternalToCertificate.java:42]: I believe it is better to try to keep this class as free as possible from logic. Maybe just pass the "IntygAvser" and let any null checks etc happen in the QuestionIntygetAvser. This removes the need to create test-cases for this particular class for handling null, which can be a bit messy when so much is going on. 

And I noticed further down that the null test is missing, so point proven ;-)

File: ts/ts-bas/src/main/java/se/inera/intyg/common/ts_bas/v7/model/converter/certificate/question/QuestionIntygetAvser.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                        )
+                    )
+                    .build()
+            )
+            .value(
+                CertificateDataValueCodeList.builder()
+                    .list(
+                        intygAvserKategori != null ? intygAvserKategori.stream()
+                            .map(QuestionIntygetAvser::getValueCode)
+                            .collect(Collectors.toList()) : List.of()
```
Comment (mww-sopra-steria) [inline ts/ts-bas/src/main/java/se/inera/intyg/common/ts_bas/v7/model/converter/certificate/question/QuestionIntygetAvser.java:121]: I would probably use Collections.emptyList() instead of List.of(). It is more clear to read... but if you leave it like this I will still approve :-)

File: ts/ts-bas/src/test/java/se/inera/intyg/common/ts_bas/v7/model/converter/certificate/question/QuestionIntygetAvserTest.java [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+            }
+
+            @Override
+            protected Integer getColumns() {
+                return 0;
+            }
+        }
+
+        @Nested
+        class IncludeValueCheckboxMultipleCodeTest extends ValueCheckboxMultipleCodeTest {
```
Comment (mww-sopra-steria) [inline ts/ts-bas/src/test/java/se/inera/intyg/common/ts_bas/v7/model/converter/certificate/question/QuestionIntygetAvserTest.java:179]: See comment on the Test-Fixture.

File: ts/ts-bas/src/test/java/se/inera/intyg/common/ts_bas/v7/model/converter/certificate/question/QuestionIntygetAvserTest.java [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+                return 0;
+            }
+        }
+    }
+
+    @Nested
+    class ToInternal {
+
+        @Test
+        void shouldIncludeKorkortsTyp() {
```
Comment (mww-sopra-steria) [inline ts/ts-bas/src/test/java/se/inera/intyg/common/ts_bas/v7/model/converter/certificate/question/QuestionIntygetAvserTest.java:245]: Any possibility to create a textFixture for managing CodeList? The challenge is similar as for the Value-text-fixture.

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
I know the col/row part is not yet finalised, but I did a review of the rest. There are a few thoughts to consider, especially on test fixtures.

File: support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java (hunk 1 above)
Comment (mww-sopra-steria) [reply support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java]: Resolved through generic testFixtures.

File: support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java (hunk 3 above)
Comment (mww-sopra-steria) [reply support/src/testFixtures/java/se/inera/intyg/common/support/facade/testsetup/model/value/ValueCheckboxMultipleCodeTest.java]: Resolved through generic testFixtures.

File: ts/ts-bas/src/test/java/se/inera/intyg/common/ts_bas/v7/model/converter/certificate/question/QuestionIntygetAvserTest.java (hunk 7 above)
Comment (mww-sopra-steria) [reply ts/ts-bas/src/test/java/se/inera/intyg/common/ts_bas/v7/model/converter/certificate/question/QuestionIntygetAvserTest.java]: Resolved through generic testFixtures.

**Review by mww-sopra-steria — APPROVED** [review]
Approved! Nice work!
